text/3374-cargo-target-features.md

```diff
@@ -0,0 +1,177 @@
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Change the `required-features` behavior
+
+* `required-features` could be changed to behave the same as `enable-features` described in this RFC (possibly over an Edition).
+However, as outlined in the [Relationship with `required-features`](#relationship-with-required-features) section, there are some use cases where the present behavior of `required-features` is desirable.
+This could also lead to a breaking change for some projects if it started building targets that were previously not included.
+* Instead of adding a separate field that lists features, a `force-enable-features = true` field could be added to change the behavior of `required-features` to have the behavior explained in this RFC.
+  That might be less confusing, but would prevent the ability to have both behaviors at the same time.
+* Only the situation where `required-features` generates an error could be changed to implicitly enable the missing features.
+  This would likely make `required-features` less annoying to work with, but doesn't help for use cases like running `cargo test` where you have specific tests or examples that you want to be automatically included (where `required-features` simply makes them silently excluded).
+* A new CLI argument could be added to change the behavior of `required-features` to behave the same as `enable-features`, avoiding the need to add `enable-features`.
+  This is viable, but the goal of this RFC is to make it as easy as possible to work with the cargo targets without requiring additional CLI arguments.
+
+## Alternate workflows
+
+* Instead of using `enable-features`, developers can be diligent in passing the appropriate `--features` options on the command-line when building their projects, possibly using `required-features` to ensure they only get built in the correct scenarios.
+  The intent of this RFC is to make that process easier and more seamless.
+* Users can set up [aliases](https://doc.rust-lang.org/cargo/reference/config.html#alias) which pass in the feature flags they want to enable.
+  This can help with a development workflow, but requires more documentation and education, and doesn't help with some commands like a remote `cargo install`.
+* Developers can organize their project in a [Cargo Workspace](https://doc.rust-lang.org/cargo/reference/workspaces.html) instead of using multiple targets within a single package.
+  This allows customizing dependencies and other settings within each package.
+  Workspaces can add some more overhead for managing multiple packages, but offer a lot more flexibility.
+  Instead of implementing this RFC, we could put more work into making workspaces easier to use, which could benefit a larger audience.
+  However, it is not clear if it is feasible to make workspaces work as effortlessly compared to targets.
+  Also, there is likely more work involved to get workspaces on the same level.
```

Comment:
If I am not mistaken, this works for binaries, but not for examples.
If you extract an example out of a package and turn it into its own stand-alone package within the same workspace you "lose" the fact that it's an example for that library/binary.

This gap is indeed an issue of its own. It leads projects with complex examples (especially ones where the `Cargo.toml` itself is an important part of the example) to move away from the example feature, so to speak.
If this gap were to be filled, I'd strongly argue that this alternative workflow would be my preference over adding more complexity to `Cargo.toml`.
It would make it easier to direct people towards workspaces or standalone packages based on their needs: start with a standalone package, if you need X you must upgrade to a workspace, which is easily done using `magical cargo command to convert into a workspace`.

Reply:
Can you say more about what you feel like you lose when the distinction of an "example" is lost? I understand that instead of running `cargo run --example foo` you have to do `cargo run -p foo`, but are there other differences? Is that mostly a perception issue?

Reply:
There is definitely a perception element.
The other difference is when they get compiled - a `cargo test` invocation will also build examples, which is quite convenient to make sure they don't rot.

Reply:
Also - as far as I'm aware - installing a crate with multiple `[[bin]]`s installs all of those bins, whereas examples are not built nor installed, and thus are purely for development/documentation/testing purposes.